crossbar/router/test/test_broker.py

```diff
@@ -511,6 +512,197 @@ class TestSession(ApplicationSession):
+                router._broker.processSubscribe(self._service_session, subscription)
+                subscription = message.Subscribe(self._service_session._session_id, u'com.example.test3')
+                router._broker.processSubscribe(self._service_session, subscription)
+
+                subscriptions = []
+                for obj in list(self._service_session._private):
+                    subscription = message.Unsubscribe(self._service_session._session_id, subscription=obj.subscription)
+                    router._broker.processUnsubscribe(self._service_session, subscription)
+                    subscriptions.append(obj.subscription)
+
+                def all_done():
+
+                    created = list(subscriptions)
+                    subscribes = list(subscriptions)
+                    unsubscribes = list(subscriptions)
+                    deletes = list(subscriptions)
+
+                    for args, argv in self._service_session._events:
+                        if args[0] == 'wamp.subscription.on_create':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_create: session id is incorrect!')
+                            test.assertTrue(args[2]['id'] in created, 'on_create: subscription id is incorrect!')
+                            created.remove(args[2]['id'])
+
+                        if args[0] == 'wamp.subscription.on_subscribe':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_subscribe: session id is incorrect!')
+                            test.assertTrue(args[2] in subscribes, 'on_subscribe: subscription id is incorrect!')
+                            subscribes.remove(args[2])
+
+                        if args[0] == 'wamp.subscription.on_unsubscribe':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_unsubscribe: session id is incorrect!')
+                            test.assertTrue(args[2] in unsubscribes, 'on_unsubscribe: subscription id is incorrect!')
+                            unsubscribes.remove(args[2])
+
+                        if args[0] == 'wamp.subscription.on_delete':
+                            test.assertEqual(args[1], self._service_session._session_id, 'on_delete: session id is incorrect!')
+                            test.assertTrue(args[2] in deletes, 'on_delete: subscription id is incorrect!')
+                            deletes.remove(args[2])
+
+                    test.assertEqual(len(created), 0, 'incorrect response sequence for on_create')
+                    test.assertEqual(len(subscribes), 0, 'incorrect response sequence for on_subscribe')
+                    test.assertEqual(len(unsubscribes), 0, 'incorrect response sequence for on_unsubscribe')
+                    test.assertEqual(len(deletes), 0, 'incorrect response sequence for on_delete')
+
+                reactor.callLater(1, all_done)
+                time.sleep(1)
+
```

Comment:
This stuff is what `Clock` is for: faking time. So we don't want to wait a real second, we want to tell the (fake) reactor: go forward in time 1 second. For example, like this: https://github.com/crossbario/autobahn-python/blob/master/autobahn/websocket/test/test_websocket.py#L183

Reply:
Yeah the sleep thing is interesting, take out the sleep and it fails .. however, I've just noticed that if I reduce the callLater to zero, and remove the sleep it works .. :)